plugin/kubernetes/controller.go

```diff
@@ -163,17 +166,40 @@ func newdnsController(ctx context.Context, kubeClient kubernetes.Interface, opts
 								h.OnAdd(d.Object)
 								dns.updateModifed()
 								recordDNSProgrammingLatency(dns.getServices(obj), apiEndpoints)
+								if !opts.skipAPIObjectsCleanup {
+									*apiEndpoints = api.Endpoints{}
+								}
 							}
 						case cache.Deleted:
+							apiEndpoints, ok := d.Object.(*api.Endpoints)
+							if !ok {
+								// Assume that the object must be a cache.DeletedFinalStateUnknown.
+								// This is essentially an indicator that the Endpoint was deleted, without a containing a
+								// up-to date copy of the Endpoints object. We need to use cache.DeletedFinalStateUnknown
+								// object so it can be properly deleted by store.Delete() below, which knows how to handle it.
+								tombstone, ok := d.Object.(cache.DeletedFinalStateUnknown)
+								if !ok {
+									return errors.New("expected tombstone")
+								}
+								if err := clientState.Delete(tombstone); err != nil {
+									return err
+								}
+								apiEndpoints, ok = tombstone.Obj.(*api.Endpoints)
```

Comment:
I don't fully understand this, why do we need this part if we already deleted it with the tombstone?

Reply:
It hasn't been deleted yet at this point. It gets deleted later with `clientState.Delete(obj)`.  If we don't call `clientState.Delete`, it will remain in the cache.

Reply:
So what does `clientState.Delete(tombstone)` do?

Reply:
arg! - I left that in by accident.  Thanks for catching that. I'll remove it and run the test again.